### PR TITLE
Add CBOR property-based tests

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Adds a set of property-based tests for System.Formats.Cbor. The tests define a model for CBOR data items and execute serialization roundtrips over randomly generated trees of the data model. The property tests have to be run manually by enabling the `CborPropertyTests` property. The data modelling has been done in F#.

Related PR #39786

cc @ViktorHofer @safern @ericstj

### CODE REVIEW

**Review by safern — APPROVED** [review]
LGTM

**Comment by danmosemsft** [comment]
@eiriktsarpalis what "area" tag should CBOR go into?

File: src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj [hunk 1]
```diff
@@ -29,4 +36,9 @@
     <Compile Include="CborRoundtripTests.cs" />
     <Compile Include="CoseKeyHelpers.cs" />
   </ItemGroup>
+
+  <ItemGroup Condition="'$(CborPropertyTests)' == 'True'">
+    <ProjectReference Include="CborDocument\System.Formats.Cbor.Tests.DataModel.fsproj" />
+    <PackageReference Include="FsCheck.Xunit" Version="2.14.3" />
```
Comment (ViktorHofer) [inline src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:42]: Please add the version to Versions.props instead.

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj [hunk 2]
```diff
@@ -2,7 +2,14 @@
   <PropertyGroup>
     <TargetFrameworks>$(NetCoreAppCurrent)</TargetFrameworks>
     <nullable>enable</nullable>
+    <CborPropertyTests>false</CborPropertyTests>
```
Comment (ViktorHofer) [inline src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:5]: Who sets this property to true?

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj (hunk 2 above)
Comment (safern) [reply src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:5]: I believe it is to have them disabled by default for now and if you want to run them you pass it globally as true.

**Comment by bartonjs** [comment]
The CBOR work was originally funded for a security project, so it has classically been area-System.Security. When we created new labels for CBOR and ASN1 there was some pushback. So, probably area-System.Security unless we found another bucket.

**Comment by danmosemsft** [comment]
Ha, yes now I remember we talked about it before. Fine. As long as we're consistent we won't bamboozle the bot.

Perhaps after we ship we may have some adjustments to do to areas eg to make System.Runtime less huge/amorphous.

File: src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:5]: Correct, I use `dotnet test -p:CborPropertyTests=true`

Comment (ViktorHofer) [reply src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:5]: Why not running this in CI?

Comment (eiriktsarpalis) [reply src/libraries/System.Formats.Cbor/tests/System.Formats.Cbor.Tests.csproj:5]: The property tests are long running (adding ~10 seconds to test runtime). So if we want to run on CI we should either put them on the outer loop or we could just reduce the number of generated inputs per property test (currently set to 10K).

**Pull request merged**